### PR TITLE
[feat]: [PL-28535]: Updating standards for API redesign

### PR DESCRIPTION
**Pull request by mankrit-singh**
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/38425)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_NA_** 
 


            
 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/38425)

**Comment by mankrit-singh** [comment]
trigger messagecheck

File: docs/api-standards/pagination.md [hunk 1]
```diff
@@ -1,12 +1,12 @@
 # Pagination
 
-For list endpoints pagination should be used with `page` and `page_size` query parameters to dictate where in the ist to get to. 
+For list endpoints pagination should be used with `page` and `limit` query parameters to dictate where in the ist to get to. 
 
-- `page_size` should default to 30
-- `page_size` should not exceed 100
+- `limit` should default to 30
```
Comment (ashishsanodia-harness) [inline docs/api-standards/pagination.md:5]: should be page

File: docs/api-standards/paths-and-actions.md [hunk 2]
```diff
@@ -93,3 +94,14 @@ use: /audit-trails/audit-trail-name
 not: /audit_trails/audit-trail-name
 not: /auditTrails/audit-trail-name
 ```
+
+## Ingress
+
+In order to expose the new APIs without redundancies in their path, we will be using Ingress rules to map our created APIs to point specifically to our endpoints.
+```
+use: https://app.harness.io/v1/orgs/{org}/projects/{project}/resource
```
Comment (ashishsanodia-harness) [inline docs/api-standards/paths-and-actions.md:102]: should be `/resources` plural in the end

**Review by TarunUba1 — APPROVED** [review]
> Reviewed 3 of 8 files at r1, 3 of 4 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38425)* status: 8 file(s) to review. 1 of 2 approvals obtained.

**Review by ashishsanodia-harness — APPROVED** [review]
> Reviewed 3 of 8 files at r1, 3 of 4 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38425)* status: :shipit: complete! ✓

File: docs/api-standards/common-parameters.md [hunk 3]
```diff
@@ -38,21 +38,18 @@
...
-
-| Field Name | Description | Default | Type | Validation |
-|------------|-------------|---------|------|------------|
-|            |             |         |      |            |
-
-
 ## Common Request Headers
+| Header Name  | Type   | Description |
+|--------------|--------|-------------|
+| Harness-Account | string | Account Identifier. This is expected to be picked up from the x-api-key, hence, it is not compulsory to pass this header parameter. |
```
Comment (allofthesepeople) [inline docs/api-standards/common-parameters.md:44]: Please remove: the account info should be retrieved from the auth token and not trusted as another input.

File: docs/api-standards/common-parameters.md [hunk 4]
```diff
@@ -38,21 +38,18 @@
...
 
 ## Common Response Headers
 
 | Header Name  | Type   | Description |
 |--------------|--------|-------------|
 | X-Request-ID | string | A short random string to corolate a request to a log message within the bounds of an account and relative time. eg urlencoded base64 of 6 bytes |
+
+## Git Experience Parameters
+
+Parameters related to Git Experience must be passed through the Request Body instead of Query Parameters, except in the case of GET calls. Going forward, these will not be picked up directly from thread context, rather, they will be explicitly populated at the Resource Layer.
```
Comment (allofthesepeople) [inline docs/api-standards/common-parameters.md:55]: This probably doesn't need called out as is typical REST approach, also see a file has been added with detail.

Additionally, an API should hide the internal detail, no need to add this here.

File: docs/api-standards/http-methods.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# HTTP Methods
```
Comment (allofthesepeople) [inline docs/api-standards/http-methods.md:1]: could this section be moved under the table here: https://github.com/harness/harness-core/blob/develop/docs/api-standards/paths-and-actions.md#default-rest-schema

File: docs/api-standards/paths-and-actions.md [hunk 6]
```diff
@@ -93,3 +94,14 @@ use: /audit-trails/audit-trail-name
 not: /audit_trails/audit-trail-name
 not: /auditTrails/audit-trail-name
 ```
+
+## Ingress
```
Comment (allofthesepeople) [inline docs/api-standards/paths-and-actions.md:98]: Implementation detail, please remove. 

However, could you possibly add a section to the README on the url of the api e.g.

```markdown
This set of documents relate to the Harness API located at `ttps://app.harness.io/v1`
```

File: docs/api-standards/paths-and-actions.md [hunk 7]
```diff
@@ -93,3 +94,14 @@ use: /audit-trails/audit-trail-name
...
+
+In order to expose the new APIs without redundancies in their path, we will be using Ingress rules to map our created APIs to point specifically to our endpoints.
+```
+use: https://app.harness.io/v1/orgs/{org}/projects/{project}/resources
+not: https://app.harness.io/gateway/v1/orgs/{org}/projects/{project}/resources
+not: https://app.harness.io/gateway/service/api/v1/orgs/{org}/projects/{project}/resources
+```
+
+For more details regarding Ingress rules, please refer to [this](https://harness.atlassian.net/wiki/spaces/PLATFORM/pages/21061108683/RFC-002+-+Remove+gateway+and+micro+service+details+from+API).
```
Comment (allofthesepeople) [inline docs/api-standards/paths-and-actions.md:107]: Again we shouldn't link to private docs from public ones :)

File: docs/api-standards/scopes-and-grouping.md [hunk 8]
```diff
@@ -0,0 +1,26 @@
+# Handling Different Scopes for the same Resource
```
Comment (allofthesepeople) [inline docs/api-standards/scopes-and-grouping.md:1]: Implementation detail on how we handle documentation please remove or move to new "spec first api development" mentioned above.

File: docs/api-standards/status-error-responses.md [hunk 9]
```diff
@@ -96,7 +96,11 @@ _Note: if using If-None-Match header for PUT actions, e.g. to get optimistic loc
 
 ## 4xx Client Error
 
-4xx range errors are to indicate the client made an error in the request and should resolve the problem before trying again. Apart from the 
+4xx range errors are to indicate the client made an error in the request and should resolve the problem before trying again. 
+
+For practically utilizing the new error handling structures, we will be adding Filters on top of our existing Exception Mappers. This ensures no conflicts with our existing processes.
```
Comment (allofthesepeople) [inline docs/api-standards/status-error-responses.md:101]: implementation detail please remove.

File: docs/api-standards/status-error-responses.md [hunk 10]
```diff
@@ -172,7 +176,7 @@ This response should be used when a rate limit is exceeded.
 
 ```json
 {
-  "error": "Unsupported Media Type"
+  "error": "Too Many Requests"
```
Comment (allofthesepeople) [inline docs/api-standards/status-error-responses.md:179]: Good catch!!

File: docs/api-standards/status-error-responses.md [hunk 11]
```diff
@@ -96,7 +96,11 @@ _Note: if using If-None-Match header for PUT actions, e.g. to get optimistic loc
 
 ## 4xx Client Error
 
-4xx range errors are to indicate the client made an error in the request and should resolve the problem before trying again. Apart from the 
+4xx range errors are to indicate the client made an error in the request and should resolve the problem before trying again. 
+
+For practically utilizing the new error handling structures, we will be adding Filters on top of our existing Exception Mappers. This ensures no conflicts with our existing processes.
+
+_Note: These error structures are still getting refined during implementation and the following structures are meant to serve as placeholders until then._
```
Comment (allofthesepeople) [inline docs/api-standards/status-error-responses.md:103]: implementation detail please remove

File: docs/api-standards/versioning.md [hunk 12]
```diff
@@ -41,6 +41,8 @@ The core principles should rarely change and marks a significant departure from
 
 Within the primary version, Harness provides a secondary level of versioning which allows for breaking changes at the service and implementation level.
 
+While the practical way to implement this kind of versioning in the backend is still being discussed, there are some general ideas around what we intend to follow:
```
Comment (allofthesepeople) [inline docs/api-standards/versioning.md:44]: implementation detail, please remove.

**Review by allofthesepeople — COMMENTED** [review]
Think I missed the merge on this but some things needing addressed: mostly as specific implementation detail and not related to the API spec.

**Pull request merged**